lib/kafka/avro_producer.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+require 'avro'
+require 'kafka/producer_manager'
+
+module Kafka
+  class AvroProducer
+    attr_reader :producer
+
+    def initialize(producer: nil)
+      @producer = producer || Kafka::ProducerManager.instance.producer
+    end
+
+    def produce(topic, payload, schema_version: 1)
+      schema = get_schema(topic, schema_version)
+      encoded_payload = encode_payload(schema, payload)
+      producer.produce_sync(topic:, payload: encoded_payload)
+    rescue => e
+      # https://karafka.io/docs/WaterDrop-Error-Handling/
+      # Errors are rescued and re-raised to demonstrate the types of errors that can occur
+      log_error(e, topic)
+      raise
+    end
+
+    private
+
+    def get_schema(topic, schema_version)
+      schema_path = Rails.root.join('lib', 'kafka', 'schemas', "#{topic}-value-#{schema_version}.avsc")
+      Avro::Schema.parse(File.read(schema_path))
+    end
+
+    def encode_payload(schema, payload)
+      validate_payload!(schema, payload)
+
+      datum_writer = Avro::IO::DatumWriter.new(schema)
+      buffer = StringIO.new
+      encoder = Avro::IO::BinaryEncoder.new(buffer)
+      datum_writer.write(payload, encoder)
+      avro_payload = buffer.string
+
+      # Add magic byte and schema ID to the payload
+      magic_byte = [0].pack('C')
+      # NOTE: This is a placeholder schema ID. In a real-world scenario, this should be fetched from a schema registry
+      # ID = 5 is the Event Bus schema ID for test schema, replace this with the actual schema ID when running locally
```

Comment:
In a real-world scenario would both the schema and the ID be fetched from the schema registry?

Reply:
Yes! I had to swap out the avro_turf gem we initially used for this purpose and that one did connect to a schema registry and performed these steps automatically. I have a followup ticket to add back that functionality with custom code.

Reply:
Perfect! It looks like I don't have permission to mark this as resolved, but it is resolved!